### ISSUE TITLE
Linear allocator doesn't re-use memory

### ISSUE THREAD

**Issue by kvark** [zakarumych/gpu-alloc#38]
Here is the relevant code:
```rust
        match &mut self.ready {
            Some(ready) if fits(self.chunk_size, ready.offset, size, align_mask) => ...,
            ready => {
                ...                
                let mut memory = device.allocate_memory(self.chunk_size, self.memory_type, flags)?;
```
This is not desired. The expectation is that both linear and buddy allocators would re-use memory.

It causes https://github.com/gfx-rs/wgpu/issues/1242 on our side.

**Comment by VincentFTS** [comment]
I still get memory allocations that shows that memory is not reused.
Have you some test code for this ?

**Comment by zakarumych** [comment]
Did you patch gpu-alloc in your dependency tree?

**Comment by zakarumych** [comment]
Made simple test where 1_000_000 linear allocations is performed, but only 1_000 is alive at a time which fit one chunk in test setup. Only two memory objects were allocated as expected.

**Comment by VincentFTS** [comment]
Strange …
Can you explain a bit how the LinearAllocator should work ?

**Comment by zakarumych** [comment]
LinearAllocator in bad english

```bad-english
For each allocation request LinearAllocator checks if active chunk can fit that request. If active chunk exists and can't fit allocation request then active chunk becomes exhausted.
If there is no active chunk (possibly because last one became exhausted) it takes pending chunk if there is one, otherwise new chunk is allocated from device.
Then it returns block of requested size from the active chunk. Block start will be end of the last block allocated from that chunk aligned up to the requested alignment (0 for first block). Numbers of block allocated from that chunk is incremented.

On deallocation chunk from which block was allocated is found by index stored in block and number of blocks allocated from it is decremented.
If number of allocated blocks from chunk becomes 0 then one of the following happens:
If this is active chunk then offset from which next allocation will be made is reset to 0, effectively reusing the chunk.
If this is exhausted chunk and there is no pending chunk then this chunk becomes pending chunk.
If this is exhausted chunk and there is pending chunk then this chunk is freed back to device.
```

The idea here is that allocations are cheap, memory overhead is minimal and with last changes two allocated memory blocks will circle between active, exhausted and pending states.
Specifically, while one chunk is exhausted, the other one is active and is used to serve block allocations.
It is expected that all blocks from exhausted chunk will be freed before active chunk is depleted. And exhausted chunk would become pending, waiting to replace active chunk.

**Comment by VincentFTS** [comment]
My bad, I only have two allocations, just as you said.
Wouldn't it be better to allocate directly two chunks making the second the prepared one ?

**Comment by zakarumych** [comment]
And what if you won't need the second one?
It is totally possible that the only one memory object will be reused.
For instance imagine game with levels. When level is loaded if allocates blocks from LinearAllocator to do uploads. If they all fit single chunk, then after level is loaded this chunk will be reused.

**Comment by VincentFTS** [comment]
Then, we may be able to reuse exhausted blocks inside the active chunk …
Here is my use case : every frame, I allocate a buffer and then deallocate the previous frame's buffer. In this situation, we always have plenty of available memory in the active chunk, but as we never reuse exhausted blocks, a second (expensive) chunk is allocated when the first chunk is filled  …

**Comment by VincentFTS** [comment]
Is there any reason why the LinearAllocator doesn't try to reuse exhausted blocks inside the active chunk ? For sure it will be harder to manage !

**Comment by zakarumych** [comment]
`FreeListAllocator` have desired block reusage property. It is now available in `gpu-alloc 0.4.4` under feature enabled by default. Feature would probably deleted in next breaking release.
